cpmpy/expressions/variables.py

```diff
@@ -468,131 +468,125 @@ def sum(self, axis=None, out=None):
         if out is not None:
             raise NotImplementedError()
 
-        if axis is None:    # simple case where we want the Minimum over the whole array
-            arr = self.flatten()
-            return Minimum(arr)
-
-        arr = self.__axis(axis=axis)
+        if axis is None:    # simple case where we want the minimum over the whole array
+            return cpm_min(self)
 
-        out = []
-        for i in range(0, arr.shape[0]):
-            out.append(Minimum(arr[i, ...]))
-
-        # return the NDVarArray that contains the Minimum global constraints
-        return out
+        return cpm_array(np.apply_along_axis(cpm_min, axis=axis, arr=self))
 
     def any(self, axis=None, out=None):
         """
             overwrite np.any(NDVarArray)
         """
-        from .python_builtins import any
+        from .python_builtins import any as cpm_any
 
-        if any(not is_boolexpr(x) for x in self.flatten()):
+        if any(not is_boolexpr(x) for x in self.flat):
             raise TypeError("Cannot call .any() in an array not consisting only of bools")
 
         if out is not None:
             raise NotImplementedError()
 
-        if axis is None:    # simple case where we want the .any() over the whole array
-            arr = self.flatten()
-            return any(arr)
-
-        arr = self.__axis(axis=axis)
+        if axis is None:    # simple case where we want a disjunction over the whole array
+            return cpm_any(self)
 
-        out = []
-        for i in range(0, arr.shape[0]):
-            out.append(any(arr[i, ...]))
+        return cpm_array(np.apply_along_axis(cpm_any, axis=axis, arr=self))
 
-        # return the NDVarArray that contains the any() constraints
-        return out
 
     def all(self, axis=None, out=None):
         """
             overwrite np.any(NDVarArray)
         """
 
-        from .python_builtins import all
+        from .python_builtins import all as cpm_all
+
+        if any(not is_boolexpr(x) for x in self.flat):
+            raise TypeError("Cannot call .any() in an array not consisting only of bools")
+
         if out is not None:
             raise NotImplementedError()
 
-        if axis is None:    # simple case where we want the .all() over the whole array
-            arr = self.flatten()
-            return all(arr)
+        if axis is None:  # simple case where we want a conjunction over the whole array
+            return cpm_all(self)
+
+        return cpm_array(np.apply_along_axis(cpm_all, axis=axis, arr=self))
+
+    def alldifferent(self, axis=None):
+        from .globalconstraints import AllDifferent
+        return cpm_array(np.apply_along_axis(lambda arr : AllDifferent(arr), axis=axis, arr=self))
+
+    def alldifferent_except0(self, axis=None):
+        from .globalconstraints import AllDifferentExcept0
+        return cpm_array(np.apply_along_axis(lambda arr: AllDifferentExcept0(arr), axis=axis, arr=self))
+
+    def allequal(self, axis=None):
```

Comment:
All of these crash when axis == None (which is the default value).

I guess they could all be valid without specifying an axis right. (would be the same as just calling the global directly)